sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSource.java

```diff
@@ -312,8 +315,14 @@ public void finalizeCheckpoint() throws IOException {
           reader.ackBatch(batchSafeToAckIds);
         }
       } finally {
-        checkState(reader.numInFlightCheckpoints.decrementAndGet() >= 0,
+        int remainingInFlight = reader.numInFlightCheckpoints.decrementAndGet();
+        checkState(remainingInFlight >= 0,
                    "Miscounted in-flight checkpoints");
+        if (!reader.active.get() && remainingInFlight == 0) {
```

Comment:
optional suggest: move all the checks into `maybeCloseClient()` and just call it unconditionally upon finalize checkpoint or reader close.